prov/verbs/src/fi_verbs.h

```diff
@@ -382,33 +431,181 @@ void fi_ibv_monitor_unsubscribe(struct ofi_mem_monitor *notifier,
+
+	fi_ibv_dgram_pool_entry_cancel_hndlr cancel_hndlr;
+};
+
+enum fi_ibv_ini_qp_state {
+	FI_IBV_INI_QP_UNCONNECTED,
+	FI_IBV_INI_QP_CONNECTING,
+	FI_IBV_INI_QP_CONNECTED
+};
+
+#define FI_IBV_NO_INI_TGT_QPNUM 0
+#define FI_IBV_RECIP_CONN	1
+
+/*
+ * An XRC transport INI QP connection can be shared within a process to
+ * communicate with all the ranks on the same remote node. This structure is
+ * only accessed during connection setup and tear down and should be
+ * done while holding the domain:xrc:ini_mgmt_lock.
+ */
+struct fi_ibv_ini_shared_conn {
+	/* To share, EP must have same remote peer host addr and TX CQ */
+	struct sockaddr			*peer_addr;
+	struct fi_ibv_cq		*tx_cq;
+
+	/* The physical INI/TGT QPN connection. Virtual connections to the
+	 * same remote peer and TGT QPN will share this connection, with
+	 * the remote end opening the specified XRC TGT QPN for sharing. */
+	enum fi_ibv_ini_qp_state	state;
+	struct ibv_qp			*ini_qp;
+	uint32_t			tgt_qpn;
+
+	/* EP waiting on or using this INI/TGT physical connection will be in
+	 * one of these list and hold a reference to the shared connection. */
+	struct dlist_entry		pending_list;
+	struct dlist_entry		active_list;
+	ofi_atomic32_t			ref_cnt;
+};
+
+enum fi_ibv_xrc_ep_conn_state {
+	FI_IBV_XRC_UNCONNECTED,
+	FI_IBV_XRC_ORIG_CONNECTING,
+	FI_IBV_XRC_ORIG_CONNECTED,
+	FI_IBV_XRC_RECIP_CONNECTING,
+	FI_IBV_XRC_CONNECTED
+};
+
+/*
+ * The following XRC state is only required during XRC connection
+ * establishment and can be freed once bidirectional connectivity
+ * is established.
+ */
+struct fi_ibv_xrc_ep_conn_setup {
+	uint32_t			tag;
+
+	/* IB CM message stale/duplicate detection processing requires
+	 * that shared INI/TGT connections use unique QP numbers during
+	 * RDMA CM connection setup. To avoid conflicts with actual HCA
+	 * QP number space, we allocate minimal QP that are left in the
+	 * reset state and closed once the setup process completes. */
+	struct ibv_qp			*rsvd_ini_qpn;
+	struct ibv_qp			*rsvd_tgt_qpn;
+
+
+	/* Delivery of the FI_CONNECTED event is delayed until
+	 * bidirectional connectivity is established. */
+	size_t				event_len;
+	uint8_t				event_data[FI_IBV_CM_DATA_SIZE];
+
+	/* Connection request may have to queue waiting for the
+	 * physical XRC INI/TGT QP connection to complete. */
+	int				pending_recip;
+	size_t				pending_paramlen;
+	uint8_t				pending_param[FI_IBV_CM_DATA_SIZE];
+};
+
 struct fi_ibv_ep {
-	struct util_ep		util_ep;
-	struct ibv_qp		*ibv_qp;
+	struct util_ep			util_ep;
+	struct ibv_qp			*ibv_qp;
 	union {
-		struct rdma_cm_id	*id;
+		struct rdma_cm_id		*id;
 		struct {
 			struct ofi_ib_ud_ep_name	ep_name;
 			int				service;
 		};
 	};
-	struct fi_ibv_eq	*eq;
-	struct fi_ibv_srq_ep	*srq_ep;
-	struct fi_info		*info;
+	struct fi_ibv_eq		*eq;
+	struct fi_ibv_srq_ep		*srq_ep;
+	struct fi_info			*info;
 
 	struct {
 		struct ibv_send_wr	rma_wr;
 		struct ibv_send_wr	msg_wr;
-		struct ibv_sge		sge;	
+		struct ibv_sge		sge;
 	} *wrs;
+
```

Comment:
Concern may be expressed about the members added to the fi_ibv_ep structure when XRC is not being used. Near term the --with-verbs-xrc allows this to not impact the RC endpoint when XRC is not configured. Long term it might be a better compromise to move the XRC only elements to a separate structure allocated only when XRC is being used. For XRC the sharing QP TX queues across endpoints more than makes up for the additional EP_MSG size (44 bytes). Just something to keep in mind going forward.